### ISSUE TITLE
missing sats after force close from peer (SCB recovery)

### ISSUE THREAD

**Issue by souspeed** [lightningnetwork/lnd#6129]
### Background

Peer's node went down and initiated an SCB recovery. This triggered a local force close on my side for our channel (4.000.000 sats). However the pending force close listed 3.600.000 as recoverable, and 400.000 sats as in limbo. Once the transaction closed on chain, I can see the 3.600.000 send to our receiving addresses (red arrow on mempool.space), and four transactions of 100.000 with a green arrow (not sure what a green arrow means). These sats should be mine, but somehow I did not receive them. 
Maybe something went wrong because they were in HTLC when the node went down.

Question:
What should I do to fix this?
And can this be fixed, to make sure the online node always receives their own sats back.

Environment
* LND 14.1
* Umbrel 4.10

### Steps to reproduce

I'm not sure how to reproduce this.

### Expected behaviour

If someone else's node goes down and they trigger an SCB recovery, then all satoshi's on my side of the channel should come back to me, even when they are in HTLC!

### Actual behaviour

Peer's node crashed and triggered SCB recovery, part of my channel's sat's remain in limbo after force close channel.

**Comment by guggero** [comment]
Without having the on-chain TX it's hard to say more. But those probably are HTLCs that were ongoing. So you just need to wait for them to time out (look at maturity height in `lncli pendingchannels`).

**Comment by souspeed** [comment]
> Without having the on-chain TX it's hard to say more. But those probably are HTLCs that were ongoing. So you just need to wait for them to time out (look at maturity height in `lncli pendingchannels`).

Ok, I will wait and check this.

**Comment by Roasbeef** [comment]
If they're HTLCs, then they can't be claimed without additional information. I'd keep watching the output of the call, and re-open this issue if you determine that these are actually normal outputs.

**Comment by souspeed** [comment]
The sats are still missing, see tx details. These are htlc's that were in progress when the peer's node went down, I think.
5919f5341050e42b1a40838c03c3022894da2a52b0b045e76f10710abcad3c53

What should I check to get more info? (lncli)

**Comment by guggero** [comment]
Hmm, are you sure that the payment you listed above caused the force-close? Because the amounts don't seem to match.

Can you please show the output of `lncli closedchannels` here?

It really depends on which direction those 4 HTLCs (4x100k sats) were extended. If they were offered from the crashed node to your node to forward, then your node can only pull them with the preimage. If the payment doesn't go through, your node doesn't have it (unless you were the ultimate recipient?) and the sending node needs to pull the funds after the timeout. But that "pull after timeout" isn't supported when you need to recover from SCB.

**Comment by souspeed** [comment]
For the pub key of this peer (02249e0dd01ffdf048c3327e6544e154bf89700912936d05587ed9a5fc4c560e16) 
I can find below info.

lncli closedchannels 

{
            "channel_point": "2313f11ee8d8c71ac7ad29968b62082b80560f3f97661f43ceb1f6303956dd2c:0",
            "chan_id": "786271760194666496",
            "chain_hash": "000000000019d6689c085ae165831e934ff763ae46a2a6c172b3f1b60a8ce26f",
            "closing_tx_hash": "5919f5341050e42b1a40838c03c3022894da2a52b0b045e76f10710abcad3c53",
            "remote_pubkey": "02249e0dd01ffdf048c3327e6544e154bf89700912936d05587ed9a5fc4c560e16",
            "capacity": "4000000",
            "close_height": 716333,
            "settled_balance": "3284932",
            "time_locked_balance": "3284932",
            "close_type": "LOCAL_FORCE_CLOSE",
            "open_initiator": "INITIATOR_REMOTE",
            "close_initiator": "INITIATOR_LOCAL",
            "resolutions": [
                {
                    "resolution_type": "ANCHOR",
                    "outcome": "CLAIMED",
                    "outpoint": {
                        "txid_bytes": "533cadbc0a71106fe745b0b0522ada942802c3038c83401a2be4501034f51959",
                        "txid_str": "5919f5341050e42b1a40838c03c3022894da2a52b0b045e76f10710abcad3c53",
                        "output_index": 1
                    },
                    "amount_sat": "330",
                    "sweep_txid": "43505d640a2ca5031b46fd92c0e88ebade7c6acb8db5e969c6e4f0369ddfd430"
                },
                {
                    "resolution_type": "COMMIT",
                    "outcome": "CLAIMED",
                    "outpoint": {
                        "txid_bytes": "533cadbc0a71106fe745b0b0522ada942802c3038c83401a2be4501034f51959",
                        "txid_str": "5919f5341050e42b1a40838c03c3022894da2a52b0b045e76f10710abcad3c53",
                        "output_index": 7
                    },
                    "amount_sat": "3284932",
                    "sweep_txid": "6d3bb1b7663567c31749738879792ca7caad89adcebc261cd9f3b88f5d4c6588"
                },
                {
                    "resolution_type": "INCOMING_HTLC",
                    "outcome": "TIMEOUT",
                    "outpoint": {
                        "txid_bytes": "650315e2a1d015b657f9cecf4c65884e3e848f8b272fac9858a0027ffa403079",
                        "txid_str": "793040fa7f02a05898ac2f278b8f843e4e88654ccfcef957b615d0a1e2150365",
                        "output_index": 0
                    },
                    "amount_sat": "100015",
                    "sweep_txid": ""
                },
                {
                    "resolution_type": "INCOMING_HTLC",
                    "outcome": "TIMEOUT",
                    "outpoint": {
                        "txid_bytes": "6b18e4c1a8bcbf96d6bbf44a49f2dc80f0a0f395a975cec03da1d47518ac6817",
                        "txid_str": "1768ac1875d4a13dc0ce75a995f3a0f080dcf2494af4bbd696bfbca8c1e4186b",
                        "output_index": 0
                    },
                    "amount_sat": "100013",
                    "sweep_txid": ""
                },
                {
                    "resolution_type": "INCOMING_HTLC",
                    "outcome": "TIMEOUT",
                    "outpoint": {
                        "txid_bytes": "b34898b5dd37888bf633a3dd44e169f1ea671c5b9d9495e55de1cbd8e4a31df8",
                        "txid_str": "f81da3e4d8cbe15de595949d5b1c67eaf169e144dda333f68b8837ddb59848b3",
                        "output_index": 0
                    },
                    "amount_sat": "100015",
                    "sweep_txid": ""
                },
                {
                    "resolution_type": "INCOMING_HTLC",
                    "outcome": "TIMEOUT",
                    "outpoint": {
                        "txid_bytes": "c594d2b57b5b4b645dd5a76ebc12a230f64f35b0f2f121aba38bab6f452c8d13",
                        "txid_str": "138d2c456fab8ba3ab21f1f2b0354ff630a212bc6ea7d55d644b5b7bb5d294c5",
                        "output_index": 0
                    },
                    "amount_sat": "100015",
                    "sweep_txid": ""
                }
            ]
        },
        {

**Comment by guggero** [comment]
Ah, so lnd did try to sweep the timed out HTLCs. But the 4 transactions never made it into the mempool (138d2c456fab8ba3ab21f1f2b0354ff630a212bc6ea7d55d644b5b7bb5d294c5, f81da3e4d8cbe15de595949d5b1c67eaf169e144dda333f68b8837ddb59848b3, 1768ac1875d4a13dc0ce75a995f3a0f080dcf2494af4bbd696bfbca8c1e4186b, 793040fa7f02a05898ac2f278b8f843e4e88654ccfcef957b615d0a1e2150365).

Can you restart lnd and check the log if any sweep transactions are attempted? And if they fail and for what reason?

**Comment by souspeed** [comment]
Restarting LND did not fix these HTLCs.
How do I check any sweep transactions in the log file? What do I need to look for?

**Comment by guggero** [comment]
Can you DM me the full log please? You can find me on [our community Slack server](https://lightning.engineering/slack.html), same username.

**Comment by guggero** [comment]
Thank you for the logs. I should have looked at the output of `lncli closedchannels` more closely. Sorry, I only noticed this when digging into the logs (too many things going on in parallel).
The 4 missing HTLCs were `INCOMING_HTLC`, meaning they were extended to your node (possibly in a forward event). So those sats belong to your peer that your node force closed on.
Your node very likely extended 4 similar HTLCs on other channels. Because the payment didn't go through, your node canceled those HTLCs or pulled them through the timeout clause if they went to chain.

So long story short: Those sats belong to the other node. Unfortunately HTLCs aren't covered by the SCB recovery as far as I'm aware.

**Comment by souspeed** [comment]
Ok, thanks for the update.
Does this mean that these HTLCs left my node, but I didn't collect the incoming HTLCs from this peer since his node failed?

I initially noticed I was missing about 400.000 sats, and after investigating I noticed these might be part of this force close from this peer.
Shouldn't an online node always be able to collect their own sats (HTLCs), regardless if another node crashes during ongoing HTLCs? 
Is this something that can be fixed for the future?

**Comment by guggero** [comment]
>  Does this mean that these HTLCs left my node, but I didn't collect the incoming HTLCs from this peer since his node failed?

No, it means those HTLCs were extended from the other (now failed) node towards your node, either as a direct payment or as part of a routing/forwarding event. That means your node could only pull them if it knew the preimage. And having the preimage would mean the payment was successful (which it doesn't appear to be, otherwise your node would've pulled them). So the payment failed and the other node would've now pulled the HTLCs through the timeout mechanism. Since that node failed and restored from SCB, it doesn't have the required information to do that anymore, as that information is not included in the SCB file.

> Is this something that can be fixed for the future?

This can already be addressed today by making sure the main channel DB isn't lost or corrupted. Things like disk mirroring/RAIDs and UPS. And in the (very near) future, the main database can also be replicated by using a remote DB like Postgres or `etcd`.

**Comment by 2Fast2BCn** [comment]
Hello, So I am this failed node.
If I somehow could revive my old node could I still recover those sats? or are they lost forever?
I misunderstood restore in thunderhub so I think I still have the channel DB file somewhere.

**Comment by guggero** [comment]
If you are sure that all other channels were closed (very important, otherwise you risk breaching yourself!), you can try starting up lnd with the old channel DB. It should then go ahead and pull the HTLCs through the timeout clause.
